lib/util.js

```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
   // Format -0 as '-0'. Strict equality won't distinguish 0 from -0.
   if (Object.is(value, -0))
-    return ctx.stylize('-0', 'number');
-  return ctx.stylize(`${value}`, 'number');
+    return fn('-0', 'number');
+  return fn(`${value}`, 'number');
 }
 
-
-function formatPrimitive(ctx, value) {
-  if (value === undefined)
-    return ctx.stylize('undefined', 'undefined');
-
-  // For some reason typeof null is "object", so special case here.
-  if (value === null)
-    return ctx.stylize('null', 'null');
-
-  var type = typeof value;
-
-  if (type === 'string') {
-    var simple = JSON.stringify(value)
-                     .replace(/^"|"$/g, '')
-                     .replace(/'/g, "\\'")
-                     .replace(/\\"/g, '"');
-    return ctx.stylize(`'${simple}'`, 'string');
-  }
+function formatPrimitive(fn, value) {
+  const type = typeof value;
+  if (type === 'string')
+    return fn(strEscape(value), 'string');
   if (type === 'number')
-    return formatNumber(ctx, value);
+    return formatNumber(fn, value);
   if (type === 'boolean')
-    return ctx.stylize(`${value}`, 'boolean');
+    return fn(`${value}`, 'boolean');
+  if (type === 'undefined')
+    return fn('undefined', 'undefined');
   // es6 symbol primitive
-  if (type === 'symbol')
-    return ctx.stylize(value.toString(), 'symbol');
-}
-
-
-function formatPrimitiveNoColor(ctx, value) {
-  var stylize = ctx.stylize;
-  ctx.stylize = stylizeNoColor;
-  var str = formatPrimitive(ctx, value);
-  ctx.stylize = stylize;
-  return str;
+  return fn(value.toString(), 'symbol');
 }
 
-
 function formatError(value) {
   return value.stack || `[${errorToString.call(value)}]`;
 }
 
-
-function formatObject(ctx, value, recurseTimes, visibleKeys, keys) {
-  return keys.map(function(key) {
-    return formatProperty(ctx, value, recurseTimes, visibleKeys, key, false);
-  });
+function formatObject(ctx, value, recurseTimes, keys) {
+  const len = keys.length;
+  const output = new Array(len);
+  for (var i = 0; i < len; i++)
+    output[i] = formatProperty(ctx, value, recurseTimes, keys[i], 0);
+  return output;
 }
 
-
-function formatArray(ctx, value, recurseTimes, visibleKeys, keys) {
-  const maxLength = Math.min(Math.max(0, ctx.maxArrayLength), value.length);
-  var output = [];
-  let visibleLength = 0;
-  let index = 0;
-  for (const elem of keys) {
-    if (visibleLength === maxLength)
-      break;
-    // Symbols might have been added to the keys
-    if (typeof elem !== 'string')
-      continue;
-    const i = +elem;
-    if (index !== i) {
-      // Skip zero and negative numbers as well as non numbers
-      if (i > 0 === false)
-        continue;
-      const emptyItems = i - index;
-      const ending = emptyItems > 1 ? 's' : '';
-      const message = `<${emptyItems} empty item${ending}>`;
-      output.push(ctx.stylize(message, 'undefined'));
-      index = i;
-      if (++visibleLength === maxLength)
+// The array is sparse and or has extra keys
```

Comment:
Minor nit: should be 'and/or' instead of 'and or'

Reply:
:+1: